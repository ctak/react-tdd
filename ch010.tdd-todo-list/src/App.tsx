import React from 'react';
import Styled from 'styled-components';
import { Button, PageHeader } from 'Components';

// import { List } from 'Pages';
// import { Add } from 'Pages';

const Container = Styled.div`
  min-height: 100vh;
  background-color: #eee;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

function App() {
  return (
    <Container>
      <PageHeader />
    </Container>
  );
}

export default App;
