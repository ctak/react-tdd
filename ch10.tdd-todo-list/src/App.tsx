import React from 'react';
import Styled from 'styled-components';
import { PageHeader } from 'Components';

const Container = Styled.div``;

function App() {
  return (
    <Container>
      <PageHeader />
    </Container>
  );
}

export default App;
