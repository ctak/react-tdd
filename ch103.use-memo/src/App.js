import React, { useRef, useState, useMemo } from 'react';

import UserList from './components/UserList';
import CreateUser from './components/CreateUser';

// let nextId = 4;

// useRef 를 사용한다는 것은, Component 가 state, props 로 다시 렌더링 될 때, function 이 다시 실행되는 것과 같으니까,
// 변수를 유지살 수 있는 방법이 필요함.
// 그렇다면 closure 처럼, 먼가 방법이 필요하겠고, 함수는 다시 실행되지만, 내부의 값은 유지되어야 한다.
// 그래서 리액트에서는 useRef() 로 closure 객체를 만든 것이다.

// useMemo 를 사용하여 연산한 값 재사용하기. WHY?
function countActiveUsers(users) {
  console.log('활성 사용자 수를 세는 중...');
  return users.filter(user => user.active).length; // return 이 Number 네.
}
function App() {
  const [inputs, setInputs] = useState({
    username: '',
    email: '',
  });
  const { username, email } = inputs;
  const onChange = e => {
    const { name, value } = e.target;
    setInputs({
      ...inputs,
      [name]: value
    });
  };
  // let nextId = 4;
  const nextId = useRef(4);
  const onCreate = () => {
    const user = {
      id: nextId.current,
      username,
      email,
    };
    setUsers([...users, user]);

    setInputs({
      username: '',
      email: '',
    });

    nextId.current += 1;
  }
  const onRemove = (id) => {
    setUsers(users.filter(user => user.id !== id));
  }
  const onToggle = (id) => {
    setUsers(
      users.map(user => (
        user.id === id
          ? { ...user, ['active']: !user.active }
          : user
      ))
    );
  }
  const [users, setUsers] = useState([
    {
      id: 1,
      username: 'velopert',
      email: 'public.velopert@gmail.com',
      active: true,
    },
    {
      id: 2,
      username: 'tester',
      email: 'tester@example.com',
      active: false,
    },
    {
      id: 3,
      username: 'liz',
      email: 'liz@example.com',
      active: false,
    },
  ]);
  // 컴포넌트에서 이런 경우는 굉장히 많겠지.
  // 이런 경우를 없애려면, useMemo()를 사용한다는 것이군.
  const count = useMemo(() => countActiveUsers(users), [users]);

  return (
    <>
      <CreateUser
        username={username}
        email={email}
        onChange={onChange}
        onCreate={onCreate}
      />
      <hr />
      <UserList
        users={users}
        onRemove={onRemove}
        onToggle={onToggle}
      />
      <div>활성 사용자 수: {count}</div>
    </>
  );
}

export default App;
