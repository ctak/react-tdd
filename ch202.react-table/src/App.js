import React from 'react';
import PharmacyListPage from './pages/PharmacyListPage';

function App() {
  return (
    <>
      <PharmacyListPage />
    </>
  );
}

export default App;
