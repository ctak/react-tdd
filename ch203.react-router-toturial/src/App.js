import { Routes, Route } from "react-router-dom";
import Contact from "./pages/Contact";
// import ErrorPage from "./pages/ErrrorPage";
import NotFound from "./pages/NotFound";
import Root from "./pages/Root";

function App() {
  return (
    <Routes>
      <Route 
        path="/" 
        element={<Root />}
        // errorElement={<ErrorPage />} // 결론은 이렇게 tutorial 처럼 나올 수 없다는 것.
      >
      </Route>
      <Route path="/contacts/:contactId" element={<Contact />} />
      <Route path="*" element={<NotFound />} />
    </Routes>
  );
}

export default App;
