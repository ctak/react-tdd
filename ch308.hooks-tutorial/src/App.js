import React, { useState } from "react";
import Average from "./components/Average";
import Counter from './components/Counter';
import CounterWithReducer from "./components/CounterUseReducer";
import Info from "./components/Info";
import InfoWithCustomHook from "./components/InfoWithCustomHook";

const App = () => {
  const [visible, setVisible] = useState(false);
  return (
    <>
      <button
        onClick={() => setVisible(!visible)}
      >
        {visible ? '숨기기' : '보이기'}
      </button>
      <hr />
      {visible && <Info />}
      <hr />
      <Counter />
      <hr />
      <CounterWithReducer />
      <hr />
      <Average />
      <hr />
      <InfoWithCustomHook />
    </>
  );
}

export default App;
