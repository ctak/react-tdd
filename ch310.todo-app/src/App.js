import { useCallback, useRef, useState } from 'react';
import TodoInsert from './components/TodoInsert';
import TodoList from './components/TodoList';
import TodoTemplate from './components/TodoTemplate';

const App = () => {
  const [todos, setTodos] = useState([
    {
      id: 1,
      text: '리액트의 기초 알아보기',
      checked: true,
    },
    {
      id: 2,
      text: '컴포넌트 스타일링해 보기',
      checked: true,
    },
    {
      id: 3,
      text: '일정 관리 앱 만들어 보기',
      checked: false,
    },
  ]);

  // 고윳값으로 사용될 id
  // ref 를 사용하여 변수 담기
  const nextId = useRef(4);

  const onInsert = useCallback(
    text => {
      const todo = {
        id: nextId.current,
        text,
        checked: false
      };

      // setTodos([
      //   ...todos,
      //   todo,
      // ]);
      setTodos(todos.concat(todo));

      nextId.current += 1;
    },
    [todos],
  )

  const onRemove = useCallback(
    id => setTodos(todos.filter(todo => todo.id !== id)),
    [todos],
  )

  const onToggle = useCallback(
    id => {
      // const nextTodos = todos.map((todo) => {
      //   if (todo.id === id) {
      //     return {
      //       ...todo,
      //       checked: !todo.checked,
      //     }
      //   } else {
      //     return todo;
      //   }
      // });
      const nextTodos = todos.map(todo =>
        todo.id === id ? { ...todo, checked: !todo.checked } : todo,
      );
      setTodos(nextTodos);
    },
    [todos],
  );

  return (
    <TodoTemplate>
      <TodoInsert onInsert={onInsert} />
      <TodoList todos={todos} onRemove={onRemove} onToggle={onToggle} />
    </TodoTemplate>
  );
};

export default App;
