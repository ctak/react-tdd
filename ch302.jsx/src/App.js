import logo from "./logo.svg";
import "./App.css";

function myFunction() {
  const a = 1;
  if (true) {
    const a = 2;
    console.log(`first: ${a}`); // maybe 2
  }
  console.log(`second: ${a}`); // maybe 1
}

function App() {
  myFunction();
  const number = 0;
  return (
    <div
      className="App"
      // color="red" // 시작 태그에서는 // 주석이 가능하다.
    >
      {number && <div>내용</div>}
    </div>
  );
}

export default App;
