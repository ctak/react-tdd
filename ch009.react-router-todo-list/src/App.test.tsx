import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';
import 'jest-styled-components';
import exp from 'constants';
import { exec } from 'child_process';

describe('<App />', () => {
  it('renders component correctly', () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Router navigator={history} location="/">
        <App />
      </Router>
    
    );

    const header = screen.getByText('할 일 목록');
    expect(header).toBeInTheDocument();

    const toDoList = screen.getByTestId('toDoList');
    expect(toDoList).toBeInTheDocument();
    expect(toDoList.firstChild).toBeNull();

    const label = screen.getByText('+');
    expect(label).toBeInTheDocument();
    
    expect(container).toMatchSnapshot();
  });

  it('goes to Add page and goBack to List page', async () => {
    const history = createMemoryHistory();
    // history.push('/');
    const { container } = render(
      <Router navigator={history} location="/">
        <App />
      </Router>
    
    );

    // @20220824 아래에 보면 왜 동일한 fireEvent.click 인데, 비동기 와 동기가 다른 것일까?

    const addButton = screen.getByText('+');
    fireEvent.click(addButton);

    setTimeout(() => {
      const header = screen.getByText('할 일 추가');
      expect(header).toBeInTheDocument();

      const goBack = screen.getByText('돌아가기');
      expect(goBack).toBeInTheDocument();

      const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
      expect(input).toBeInTheDocument();

      const button = screen.getByText('추가');
      expect(button).toBeInTheDocument();

      expect(container).toMatchSnapshot();

      fireEvent.click(goBack);
      expect(header.textContent).toBe('할 일 목록');
      const toDoList = screen.getByTestId('toDoList');
      expect(toDoList).toBeInTheDocument();
    });
    
  })

  it('goes to Detail page and go back to List page', () => {
    localStorage.setItem('ToDoList', '["ToDo 1"]');

    const history = createMemoryHistory();

    const { container } = render(
      <Router navigator={history} location="/">
        <App />
      </Router>
    );

    const toDoItem = screen.getByText('ToDo 1');
    expect(toDoItem).toBeInTheDocument();
    fireEvent.click(toDoItem);

    setTimeout(() => {
      const header = screen.getByText('할 일 상세');
      expect(header).toBeInTheDocument();
      const goBack = screen.getByText('돌아가기');
      expect(goBack).toBeInTheDocument();
      const toDo = screen.getByText('ToDo 1');
      expect(goBack).toBeInTheDocument();
      const button = screen.getByText('삭제');
      expect(button).toBeInTheDocument();

      expect(container).toMatchSnapshot();

      fireEvent.click(goBack);
      expect(header.textContent).toBe('할 일 목록');
      const toDoList = screen.getByTestId('toDoList');
      expect(toDoList).toBeInTheDocument();
    })
    
  })

  it('shows Not Found page if the user enters the wrong URL, and go back to List page', () => {
    const history = createMemoryHistory();

    const { container } = render(
      <Router navigator={history} location="/foo">
        <App />
      </Router>
    );

    const header = screen.getByText('에러');
    expect(header).toBeInTheDocument();
    const goBack = screen.getByText('돌아가기');
    expect(goBack).toBeInTheDocument();
    const notFoundMessage = screen.getByText('Not Found');
    expect(notFoundMessage).toBeInTheDocument();

    expect(container).toMatchSnapshot();

    fireEvent.click(goBack);
    setTimeout(() => {
      expect(header.textContent).toBe('할 일 목록');
      const toDoList = screen.getByTestId('toDoList');
      expect(toDoList).toBeInTheDocument();
    })

  })

  it('adds a new ToDo', () => {
    const history = createMemoryHistory();
    render(
      <Router navigator={history} location={"/"}>
        <App />
      </Router>
    );

    const addButton = screen.getByText('+');
    fireEvent.click(addButton);

    setTimeout(() => {
      const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
      const button = screen.getByText('추가');
      fireEvent.change(input, { target: { value: 'New ToDo' } });
      fireEvent.click(button);

      const header = screen.getByText('할 일 목록');
      expect(header).toBeInTheDocument();
      const newToDo = screen.getByText('New ToDo');
      expect(newToDo).toBeInTheDocument();
    });

  })

  it('deletes ToDo from ToDo List page', () => {
    localStorage.setItem('ToDoList', '["ToDo 1"]');

    const history = createMemoryHistory();

    render(
      <Router location={"/"} navigator={history}>
        <App />
      </Router>
    );

    const toDoItem = screen.getByText('ToDo 1');
    const deleteButton = screen.getByText('삭제');
    expect(toDoItem).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();

    fireEvent.click(deleteButton);
    expect(toDoItem).not.toBeInTheDocument();
    expect(deleteButton).not.toBeInTheDocument();
    expect(localStorage.getItem('ToDoList') as string).toBe('[]');
  })

  it('deletes ToDo from teh detail page', () => {
    localStorage.setItem('ToDoList', '["ToDo 1"]');

    const history = createMemoryHistory();
    
    render(
      <Router location={"/"} navigator={history}>
        <App />
      </Router>
    );

    const toDoItem = screen.getByText('ToDo 1');
    expect(toDoItem).toBeInTheDocument();

    fireEvent.click(toDoItem);

    setTimeout(() => {
      const header = screen.getByText('할 일 상세');
      expect(header).toBeInTheDocument();

      const deleteButton = screen.getByText('삭제');
      fireEvent.click(deleteButton);

      // 역시나 로드 된 것은 바로 찾아간다. 비동기가 아니라는 것이지!
      expect(header.textContent).toBe('할 일 목록');
      const toDoList = screen.getByTestId('toDoList');
      expect(toDoList).toBeInTheDocument();
      expect(toDoList.firstChild).toBeNull();
      expect(localStorage.getItem('ToDoList')).toBe('[]');
    })
  })

  /*
  it('adds and deletes ToDo items', () => {
    render(<App />);

    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
    const button = screen.getByText('추가');
    fireEvent.change(input, { target: { value: 'study react 1' } });
    fireEvent.click(button);

    const todoItem = screen.getByText('study react 1');
    expect(todoItem).toBeInTheDocument();
    const deleteButton = screen.getByText('삭제');
    expect(deleteButton).toBeInTheDocument();

    const toDoList = screen.getByTestId('toDoList');
    expect(toDoList.childElementCount).toBe(1);

    fireEvent.change(input, { target: { value: 'study react 2' } });
    fireEvent.click(button);

    const todoItem2 = screen.getByText('study react 2');
    expect(todoItem2).toBeInTheDocument();
    expect(toDoList.childElementCount).toBe(2);

    const deleteButtons = screen.getAllByText('삭제');
    fireEvent.click(deleteButtons[0]);

    expect(todoItem).not.toBeInTheDocument();
    expect(toDoList.childElementCount).toBe(1);
  });

  it('does not add empty ToDo', () => {
    render(<App />);

    const toDoList = screen.getByTestId('toDoList');
    const length = toDoList.childElementCount;

    const button = screen.getByText('추가');
    fireEvent.click(button);

    expect(toDoList.childElementCount).toBe(length);
  });

  it('loads localStorage data', () => {
    localStorage.setItem('ToDoList','["ToDo 1", "ToDo 2", "ToDo 3"]');
    render(<App />);

    expect(screen.getByText('ToDo 1')).toBeInTheDocument();
    expect(screen.getByText('ToDo 2')).toBeInTheDocument();
    expect(screen.getByText('ToDo 3')).toBeInTheDocument();
    expect(screen.getAllByText('삭제').length).toBe(3);
  })
  */
});