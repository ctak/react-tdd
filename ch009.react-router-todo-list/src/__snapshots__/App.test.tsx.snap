// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders component correctly 1`] = `
.c4 {
  min-height: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
  min-width: 300px;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #1e40ff;
}

.c2 {
  padding: 20px;
  color: #fff;
  font-size: 20px;
  font-weight: 600;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  font-size: 20px;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 60px;
  height: 60px;
  border-radius: 30px;
  cursor: pointer;
  position: absolute;
  bottom: -30px;
  background-color: #304ffe;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5:hover {
  background-color: #1e40ff;
}

.c5:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c0 {
  min-height: 100vh;
  background-color: #eee;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        할 일 목록
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
        data-testid="toDoList"
      />
      <a
        class="c5"
        href="/add"
      >
        +
      </a>
    </div>
  </div>
</div>
`;

exports[`<App /> shows Not Found page if the user enters the wrong URL, and go back to List page 1`] = `
.c1 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #1e40ff;
}

.c2 {
  padding: 20px;
  color: #fff;
  font-size: 20px;
  font-weight: 600;
}

.c3 {
  padding: 20px;
  color: #fff;
  font-size: 16px;
  font-weight: 600;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: absolute;
  left: 20px;
}

.c4 {
  font-size: 20px;
}

.c0 {
  min-height: 100vh;
  background-color: #eee;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        에러
      </div>
      <a
        class="c3"
        href="/"
      >
        돌아가기
      </a>
    </div>
    <div
      class="c4"
    >
      Not Found
    </div>
  </div>
</div>
`;
